pymodbus/client/base.py

```diff
@@ -121,13 +121,17 @@ async def async_execute(self, request) -> ModbusResponse:
                 except asyncio.exceptions.TimeoutError:
                     count += 1
         if count > self.retries:
-            self.ctx.connection_lost(asyncio.TimeoutError("Server not responding"))
-            raise ModbusIOException(
-                f"ERROR: No response received after {self.retries} retries"
-            )
+            self.no_message_response_after_retries(self.retries)
 
         return resp  # type: ignore[return-value]
 
+    def no_message_response_after_retries(self, retries) -> None:
```

Comment:
Is there a reason to single out tcp ? tls and udp also use socket, so they should at the very least be included, and I cannot see it hurts to include serial as well, and thus make a general solution (which is much less confusing).

Reply:
I was being cautious and not changing behaviour I can't test. The thing about TCP is that we do have a connection protocol and hence the lower layers will tell us if the connection needs to be closed and reopened. (i.e. there's a FIN or RST packet, or a request is not ack'ed at the transport layer.) This isn't true of UDP - though it is for TLS. 

Happy to be guided by you though - many thanks for comments.

Reply:
This is very true for UDP as well, in the way we use it.

For serial it is the same since we have the device drivers to tell us if there are problems.

We do not want special solutions for different transport, that confuses the end users, and makes maintenance more difficult.